// @sinclair/hammer/options/options.js
"use strict";
/*--------------------------------------------------------------------------

MIT License

Copyright (c) Hammer 2021 Haydn Paterson (sinclair) <haydn.developer@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

---------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.parse = exports.parseTaskOptions = exports.parseMonitorOptions = exports.parseServeOptions = exports.parseRunOptions = exports.parseWatchOptions = exports.parseBuildOptions = exports.OptionsError = void 0;
const fs = require("fs");
const path = require("path");
// ------------------------------------------------------------------------
// Error
// ------------------------------------------------------------------------
class OptionsError extends Error {
    constructor(option, reason) {
        super(`${option}: ${reason}`);
        this.option = option;
        this.reason = reason;
    }
}
exports.OptionsError = OptionsError;
// ------------------------------------------------------------------------
// Defaults
// ------------------------------------------------------------------------
function defaultHelpOptions(message) {
    return { type: 'help', message };
}
function defaultVersionOptions() {
    return { type: 'version', version: 'resolve from package.json' };
}
function defaultBuildOptions() {
    return {
        type: 'build',
        dist: path.join(process.cwd(), 'dist'),
        esm: false,
        minify: false,
        platform: 'browser',
        sourcePaths: [],
        sourcemap: false,
        target: ['esnext'],
        external: []
    };
}
function defaultWatchOptions() {
    return {
        type: 'watch',
        dist: path.join(process.cwd(), 'dist'),
        esm: false,
        minify: false,
        platform: 'browser',
        sourcePaths: [],
        sourcemap: false,
        target: ['esnext'],
        external: []
    };
}
function defaultServeOptions() {
    return {
        type: 'serve',
        sourcePaths: [],
        dist: path.join(process.cwd(), 'dist'),
        esm: false,
        target: ['esnext'],
        port: 5000,
        sourcemap: false,
        minify: false,
        external: [],
        cors: false,
        sabs: false
    };
}
function defaultRunOptions() {
    return {
        type: 'start',
        sourcePath: 'index.ts',
        entryPath: path.join(process.cwd(), 'dist', 'index.js'),
        args: [],
        dist: path.join(process.cwd(), 'dist'),
        esm: false,
        target: ['esnext'],
        sourcemap: false,
        minify: false,
        external: []
    };
}
function defaultMonitorOptions() {
    return {
        type: 'monitor',
        sourcePaths: [],
        arguments: []
    };
}
function defaultTaskOptions() {
    return {
        type: 'task',
        sourcePath: path.join(process.cwd(), 'hammer.mjs'),
        name: '',
        arguments: []
    };
}
// ------------------------------------------------------------------------
// Preparation
// ------------------------------------------------------------------------
function prepareArguments(args) {
    let open = false;
    const result = [];
    const temp = [];
    while (args.length > 0) {
        const next = args.shift();
        if (!open && next.indexOf('"') === 0 && next.lastIndexOf('"') !== next.length - 1) {
            temp.push(next.slice(1));
            open = true;
        }
        else if (open && next.indexOf('"') === next.length - 1) {
            temp.push(next.slice(0, next.length - 1));
            result.push(temp.join(' '));
            while (temp.length > 0)
                temp.shift();
            open = false;
        }
        else if (open) {
            temp.push(next);
        }
        else {
            result.push(next);
        }
    }
    return result;
}
// ------------------------------------------------------------------------
// Field Parsers
// ------------------------------------------------------------------------
function* parseSourcePaths(params) {
    if (params.length === 0)
        throw new OptionsError('paths', 'Expected one or more paths.');
    const next = params.shift();
    for (const partialPath of next.split(' ')) {
        const sourcePath = path.resolve(process.cwd(), partialPath);
        if (!fs.existsSync(sourcePath))
            throw new OptionsError('paths', `Cannot find source path '${sourcePath} does not exists.'`);
        yield sourcePath;
    }
}
function parseSourcePathAndArguments(params) {
    if (params.length === 0)
        throw new OptionsError('run', 'Expected entry path.');
    const next = params.shift();
    const split = next.split(' ');
    const partialPath = split.shift();
    const entryPath = path.join(process.cwd(), partialPath);
    const extname = path.extname(entryPath);
    if (!['.ts', '.tsx', '.js'].includes(extname))
        throw new OptionsError('run', `Entry path not a TypeScript or JavaScript file. Got '${entryPath}'`);
    if (!fs.existsSync(entryPath))
        throw new OptionsError('run', `Entry path '${entryPath}' does not exist.`);
    return [entryPath, split];
}
function parseDist(params) {
    if (params.length === 0)
        throw new OptionsError('dist', "Expected directory path.");
    return path.join(process.cwd(), params.shift());
}
function parsePlatform(params) {
    if (params.length === 0)
        throw new OptionsError('platform', "Expected profile option.");
    const next = params.shift();
    if (next === 'browser' || next === 'node')
        return next;
    throw new OptionsError('platform', `Expected 'node' or 'browser'. Got '${next}'`);
}
function parseExternal(params) {
    if (params.length === 0)
        throw new OptionsError('external', "Expected external option.");
    let next = params.shift();
    if (next.indexOf('"') === 0)
        next = next.slice(1, next.length - 1);
    if (next.lastIndexOf('"') === next.length - 1)
        next = next.slice(0, next.length - 1);
    const external = next.split(' ').map(x => x.trim());
    return external;
}
function parsePort(params) {
    if (params.length === 0)
        throw new OptionsError('port', "Expected port number");
    const next = params.shift();
    const port = parseInt(next);
    if (Number.isNaN(port))
        throw new OptionsError('port', `Expected port number. Received '${port}'`);
    if (port < 0 || port > 65535)
        throw new OptionsError('port', `Invalid port number. Received '${port}'`);
    return port;
}
function* parseTarget(params) {
    if (params.length === 0)
        throw new OptionsError('target', "Expected target option");
    while (params.length > 0) {
        const next = params.shift();
        if (next.includes('--')) {
            params.unshift(next);
            return;
        }
        yield next;
    }
}
// ------------------------------------------------------------------------
// Parsers
// ------------------------------------------------------------------------
function parseBuildOptions(params) {
    const options = defaultBuildOptions();
    options.sourcePaths = [...parseSourcePaths(params)];
    if (options.sourcePaths.length === 0)
        throw new OptionsError('build', `Expected at least one source path.`);
    while (params.length > 0) {
        const next = params.shift();
        switch (next) {
            case '--platform':
                options.platform = parsePlatform(params);
                break;
            case '--target':
                options.target = [...parseTarget(params)];
                break;
            case '--dist':
                options.dist = parseDist(params);
                break;
            case '--esm':
                options.esm = true;
                break;
            case '--external':
                options.external = parseExternal(params);
                break;
            case '--sourcemap':
                options.sourcemap = true;
                break;
            case '--minify':
                options.minify = true;
                break;
        }
    }
    return options;
}
exports.parseBuildOptions = parseBuildOptions;
function parseWatchOptions(params) {
    const options = defaultWatchOptions();
    options.sourcePaths = [...parseSourcePaths(params)];
    if (options.sourcePaths.length === 0)
        throw new OptionsError('build', `Expected at least one source path.`);
    while (params.length > 0) {
        const next = params.shift();
        switch (next) {
            case '--platform':
                options.platform = parsePlatform(params);
                break;
            case '--target':
                options.target = [...parseTarget(params)];
                break;
            case '--dist':
                options.dist = parseDist(params);
                break;
            case '--esm':
                options.esm = true;
                break;
            case '--external':
                options.external = parseExternal(params);
                break;
            case '--sourcemap':
                options.sourcemap = true;
                break;
            case '--minify':
                options.minify = true;
                break;
        }
    }
    return options;
}
exports.parseWatchOptions = parseWatchOptions;
function parseRunOptions(params) {
    const options = defaultRunOptions();
    const [sourcePath, args] = parseSourcePathAndArguments(params);
    options.sourcePath = sourcePath;
    options.args = args;
    while (params.length > 0) {
        const next = params.shift();
        switch (next) {
            case '--dist':
                options.dist = parseDist(params);
                break;
            case '--esm':
                options.esm = true;
                break;
            case '--external':
                options.external = parseExternal(params);
                break;
            case '--target':
                options.target = [...parseTarget(params)];
                break;
            case '--sourcemap':
                options.sourcemap = true;
                break;
            case '--minify':
                options.minify = true;
                break;
        }
    }
    const extension = path.extname(options.sourcePath);
    options.entryPath = path.join(options.dist, [path.basename(options.sourcePath, extension), '.js'].join(''));
    return options;
}
exports.parseRunOptions = parseRunOptions;
function parseServeOptions(params) {
    const options = defaultServeOptions();
    options.sourcePaths = [...parseSourcePaths(params)];
    while (params.length > 0) {
        const next = params.shift();
        switch (next) {
            case '--dist':
                options.dist = parseDist(params);
                break;
            case '--esm':
                options.esm = true;
                break;
            case '--external':
                options.external = parseExternal(params);
                break;
            case '--minify':
                options.minify = true;
                break;
            case '--sourcemap':
                options.sourcemap = true;
                break;
            case '--target':
                options.target = [...parseTarget(params)];
                break;
            case '--port':
                options.port = parsePort(params);
                break;
            case '--cors': options.cors = true;
            case '--sabs': options.sabs = true;
        }
    }
    return options;
}
exports.parseServeOptions = parseServeOptions;
function parseMonitorOptions(params) {
    const options = defaultMonitorOptions();
    options.sourcePaths = [...parseSourcePaths(params)];
    options.arguments = params;
    return options;
}
exports.parseMonitorOptions = parseMonitorOptions;
function resolveHammerTaskFile() {
    if (fs.existsSync(path.join(process.cwd(), 'hammer.mjs'))) {
        return path.join(process.cwd(), 'hammer.mjs');
    }
    else {
        throw Error(`No task file found. Expected 'hammer.mjs' in current directory.`);
    }
}
function parseTaskOptions(params) {
    const options = defaultTaskOptions();
    options.sourcePath = resolveHammerTaskFile();
    if (params.length === 0)
        return options;
    options.name = params.shift();
    options.arguments = params;
    return options;
}
exports.parseTaskOptions = parseTaskOptions;
function parse(args) {
    try {
        const params = prepareArguments(args).slice(2);
        if (params.length === 0)
            return defaultHelpOptions();
        const command = params.shift();
        switch (command) {
            case 'build': return parseBuildOptions(params);
            case 'watch': return parseWatchOptions(params);
            case 'serve': return parseServeOptions(params);
            case 'run': return parseRunOptions(params);
            case 'monitor': return parseMonitorOptions(params);
            case 'task': return parseTaskOptions(params);
            case 'help': return defaultHelpOptions();
            case 'version': return defaultVersionOptions();
            default: return defaultHelpOptions(`Unknown command '${command}'`);
        }
    }
    catch (error) {
        if (error instanceof Error)
            return defaultHelpOptions(error.message);
        else {
            return defaultHelpOptions(error);
        }
    }
}
exports.parse = parse;
