// @sinclair/hammer/async/channel.js
"use strict";
/*--------------------------------------------------------------------------

MIT License

Copyright (c) Hammer 2021 Haydn Paterson (sinclair) <haydn.developer@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

---------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.select = exports.channel = exports.Receiver = exports.Sender = exports.Queue = exports.defer = void 0;
function defer() {
    let resolver;
    let rejector;
    const promise = new Promise((resolve, reject) => {
        resolver = resolve;
        rejector = reject;
    });
    return [promise, resolver, rejector];
}
exports.defer = defer;
// --------------------------------------------------------------------------
// Queue
// --------------------------------------------------------------------------
class Queue {
    constructor() {
        this.promises = [];
        this.resolvers = [];
    }
    dequeue() {
        if (this.promises.length > 0) {
            const promise = this.promises.shift();
            return promise;
        }
        else {
            const [promise, resolver] = defer();
            this.resolvers.push(resolver);
            return promise;
        }
    }
    enqueue(value) {
        if (this.resolvers.length > 0) {
            const resolver = this.resolvers.shift();
            resolver(value);
        }
        else {
            const [promise, awaiter] = defer();
            awaiter(value);
            this.promises.push(promise);
        }
    }
}
exports.Queue = Queue;
// --------------------------------------------------------------------------
// Sender
// --------------------------------------------------------------------------
class Sender {
    constructor(queue) {
        this.queue = queue;
    }
    send(value) {
        this.queue.enqueue({ type: 'data', value });
    }
    end() {
        this.queue.enqueue({ type: 'end' });
    }
}
exports.Sender = Sender;
// --------------------------------------------------------------------------
// Receiver
// --------------------------------------------------------------------------
class Receiver {
    constructor(queue) {
        this.queue = queue;
    }
    async *[Symbol.asyncIterator]() {
        while (true) {
            const next = await this.queue.dequeue();
            switch (next.type) {
                case 'data':
                    yield next.value;
                    break;
                case 'end':
                    return;
            }
        }
    }
}
exports.Receiver = Receiver;
function channel() {
    const queue = new Queue();
    const sender = new Sender(queue);
    const receiver = new Receiver(queue);
    return [sender, receiver];
}
exports.channel = channel;
function select(receivers) {
    async function receive(sender, receiver) {
        for await (const value of receiver) {
            await sender.send(value);
        }
        await sender.end();
    }
    const [sender, receiver] = channel();
    const promises = receivers.map(source => receive(sender, source));
    Promise.all(promises).then(() => sender.end());
    return receiver;
}
exports.select = select;
